server/server.go

```diff
@@ -1147,3 +1156,180 @@ func (s *Server) getClientConnectURLs() []string {
+	}
+	return hostports
+}
+
+// Ports describes URLs that the server can be contacted in
+type Ports struct {
+	Nats       []string `json:"nats,omitempty"`
+	Monitoring []string `json:"monitoring,omitempty"`
+	Cluster    []string `json:"cluster,omitempty"`
+	Profile    []string `json:"profile,omitempty"`
+}
+
+// Returns a Ports struct describing ports where the server can be contacted
+func (s *Server) PortsInfo(opts *Options) Ports {
+	s.mu.Lock()
+	info := s.copyInfo()
+	listener := s.listener
+	httpListener := s.http
+	clusterListener := s.routeListener
+	profileListener := s.profiler
+	s.mu.Unlock()
+
+	ports := Ports{}
+
+	if listener != nil {
+		natsProto := "nats"
+		if info.TLSRequired {
+			natsProto = "tls"
+		}
+		ports.Nats = formatURL(natsProto, listener)
+	}
+
+	if httpListener != nil {
+		monProto := "http"
+		if opts.HTTPSPort != 0 {
+			monProto = "https"
+		}
+		ports.Monitoring = formatURL(monProto, httpListener)
+	}
+
+	if clusterListener != nil {
+		clusterProto := "nats"
+		if opts.Cluster.TLSConfig != nil {
+			clusterProto = "tls"
+		}
+		ports.Cluster = formatURL(clusterProto, clusterListener)
+	}
+
+	if profileListener != nil {
+		ports.Profile = formatURL("http", profileListener)
+	}
+
+	return ports
+}
+
+// Returns the portsFile or empty string if not set
+func (s *Server) PortsFile(dirname string) string {
+	if dirname == _EMPTY_ {
+		return _EMPTY_
+	}
+	return path.Join(dirname, fmt.Sprintf("%s_%d.ports", path.Base(os.Args[0]), os.Getpid()))
+}
+
+func (s *Server) deletePortsFile(dirname string) {
+	portsFile := s.PortsFile(dirname)
+	if portsFile != "" {
+		if err := os.Remove(portsFile); err != nil {
+			s.Errorf("Error cleaning up ports file %s: %v", portsFile, err)
+		}
+	}
+}
+
+// Writes a file with a serialized Ports to the specified ports_file_dir.
+// The name of the file is `exename_pid.ports`, typically gnatsd_pid.ports.
+// if ports file is not set, this function has no effect
+func (s *Server) logPorts() {
+	opts := s.getOpts()
+	portsFile := s.PortsFile(opts.PortsFileDir)
+	if portsFile != _EMPTY_ {
+		go func() {
+			if s.ReadyForListeners(opts, 5*time.Second) {
+				info := s.PortsInfo(opts)
+				data, err := json.Marshal(info)
+				if err != nil {
+					s.Errorf("Error marshaling ports file: %v", err)
+					return
+				}
+				if err := ioutil.WriteFile(portsFile, data, 0666); err != nil {
+					s.Errorf("Error writing ports file (%s): %v", portsFile, err)
+					return
+				}
+			}
+		}()
+	}
+}
+
+// waits until a calculated list of listeners is resolved or a timeout
+func (s *Server) ReadyForListeners(opts *Options, dur time.Duration) bool {
+	// this to allow external
+	if opts == nil {
+		opts = s.getOpts()
+	}
+	end := time.Now().Add(dur)
+	for time.Now().Before(end) {
+		s.mu.Lock()
+		shuttingDown := s.shutdown
+		listeners := s.serviceListeners(opts)
+		s.mu.Unlock()
+
+		if shuttingDown {
+			return false
+		}
+		ok := true
+		for _, l := range listeners {
+			if ok {
```

Comment:
No big deal, but you are iterating over all listeners even when one is not ready. I would do:
```
if l == nil {
    ok = false
    break
}
```